Post Editor: use the post type singular_name as the root Block Breadcrumb label

ramonjd opened the pull request:
## Description
https://github.com/WordPress/gutenberg/pull/17311 introduced post type labels as the document label in the sidebar.

This PR does the same trick, but for the Block Breadcrumb.

Context: https://github.com/WordPress/gutenberg/pull/32528#issuecomment-857375548

This PR only affects the post editor. If a post type is unknown or cannot be found, we default back to `__( 'Document' )`.

## How has this been tested?

1. Open a post. Check that the sidebar document label is "Post" and matches the Block Breadcrumb root label, which should also be "Post"
2. Open a page. Check that the sidebar document label is "Page" and matches the Block Breadcrumb root label, which should also be "Page"


Unit test for the selector:
```js
npm run test-unit packages/editor/src/store/test/selectors.js
```

## Screenshots <!-- if applicable -->
<img width="866" alt="Screen Shot 2021-06-11 at 3 53 44 pm" src="https://user-images.githubusercontent.com/6458278/121638244-fc019280-cacd-11eb-886d-075a2828ac43.png">
<img width="1041" alt="Screen Shot 2021-06-11 at 3 54 15 pm" src="https://user-images.githubusercontent.com/6458278/121638252-ff951980-cacd-11eb-9ee9-d5454b8ceb6a.png">


## Types of changes
Adds the post type label to the Block Breadcrumb root label.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

ramonjd committed:
Fetching the post type label from the store and using it as the Breadcrumb root label.

ramonjd committed:
Moving logic into a selector

ramonjd committed:
Adding tests

kevin940726 commented on packages/editor/src/store/selectors.js:
```diff
@@ -1748,3 +1748,21 @@ export function __experimentalGetTemplateInfo( state, template ) {
+		// Disable reason: Post type labels object is shaped like this.
+		// eslint-disable-next-line camelcase
+		return postType?.labels?.singular_name || null;
+		// translators: Default label for the Document sidebar tab, not selected.
```
Forgot to remove 😆?

ramonjd replied:
> Forgot to remove 😆?

Let's keep this between you and me 😆 

P.S. Thanks for spotting!

kevin940726 commented on packages/edit-post/src/components/sidebar/settings-header/index.js:
```diff
@@ -19,16 +18,10 @@ const SettingsHeader = ( { sidebarName } ) => {
-				postType?.labels?.singular_name ??
-				// translators: Default label for the Document sidebar tab, not selected.
-				__( 'Document' ),
+			documentLabel: postTypeLabel || __( 'Document' ),
```
Should move the comments here

kevin940726 reviewed (approved):
Code LGTM! Thank you! ❤️

ramonjd committed:
Adding missing translators comments
Removing unnecessary comment in the selector.

ntsekouras commented on packages/edit-post/src/components/layout/index.js:
```diff
@@ -123,6 +127,8 @@ function Layout( { styles } ) {
 				'showBlockBreadcrumbs'
 			),
+			// translators: Default label for the Document in the Block Breadcrumb.
+			documentLabel: postTypeLabel || __( 'Document' ),
```
I guess you could use `_x` here.

ramonjd replied:
Good call. Added to be more explicit about the word class. 👍

ntsekouras commented on packages/editor/src/store/selectors.js:
```diff
@@ -1748,3 +1748,20 @@ export function __experimentalGetTemplateInfo( state, template ) {
+		const postType = select( coreStore ).getPostType( currentPostType );
+		// Disable reason: Post type labels object is shaped like this.
+		// eslint-disable-next-line camelcase
+		return postType?.labels?.singular_name || null;
```
Is there a reason for explicitly returning `null` here? We can remove this, no?

ramonjd replied:
Yes, that's me just being a control freak and being explicit. 

Optional chaining will return `undefined` anyway.

Removed ✅

ntsekouras reviewed (commented):
Should we also update the root when editing a template? I guess it's okay for a follow up though.

ramonjd committed:
Relying on optional chaining returning undefined rather than explicitly returning null in the `getPostTypeLabel` selector.
Using `_x` so we can add context for the word Document, so we can differentiate between noun/verb.

Updated tests

ramonjd commented:
> Should we also update the root when editing a template? I guess it's okay for a follow up though.

Definitely! 

I did want to add it to this PR however templates/template parts don't return a `postType` so I thought it might be neater to do the investigation/required work in a separate PR.

ntsekouras commented on packages/edit-post/src/components/layout/index.js:
```diff
@@ -123,6 +127,8 @@ function Layout( { styles } ) {
 				'showBlockBreadcrumbs'
 			),
+			// translators: Default label for the Document in the Block Breadcrumb.
+			documentLabel: postTypeLabel || _x( 'Document', 'noun' ),
```
Actually you'll have to remove the translators comment and add it as a second argument. Example here: https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/unit-control/utils.js#L126

ramonjd replied:
Is that right?

I always thought context was rather for disambiguation between strings. See [docs](https://codex.wordpress.org/I18n_for_WordPress_Developers#Disambiguation_by_context)

Translators comments provide an explanation for the string.

ntsekouras replied:
Hm.. after reading more about it, it seems it can probably work better in this case. Thank you for pointing it out!

ntsekouras reviewed (approved):
Looks good! Let's 🚢  and iterate on template title.

Logic could probably be like in the `TemplateTitle` [component](https://github.com/WordPress/gutenberg/blob/trunk/packages/edit-post/src/components/header/template-title/index.js#L18).

ramonjd commented:
> Logic could probably be like in the TemplateTitle component.

Thank you for the thorough review and for the tip about TemplateTitle 🙇 !!
